src/transform/src/will_distinct.rs

```diff
@@ -0,0 +1,144 @@
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Transform that pushes down the information that a collection will be subjected to a `Distinct` on specific columns.
+
+use mz_expr::{MirRelationExpr, MirScalarExpr};
+
+use crate::analysis::{DerivedView, NonNegative};
+
+use crate::{TransformCtx, TransformError};
+
+/// Pushes down the information that a collection will be subjected to a `Distinct` on specific columns.
+///
+/// This intends to recognize idiomatic stacks of `UNION` from SQL, which look like `Distinct` over `Union`, potentially
+/// over other `Distinct` expressions. It is only able to see patterns of `DISTINCT UNION DISTINCT, ..` and other operators
+/// in between will prevent the necessary insight to remove the second `DISTINCT`.
+///
+/// There are other potential optimizations, for example fusing multiple reads of the same source, where distinctness may
+/// mean they could be a single read (with additional requirements).
+#[derive(Debug, Default)]
+pub struct WillDistinct;
+
+impl crate::Transform for WillDistinct {
+    #[mz_ore::instrument(
+        target = "optimizer"
+        level = "trace",
+        fields(path.segment = "will_distinct")
+    )]
+    fn transform(
+        &self,
+        relation: &mut MirRelationExpr,
+        ctx: &mut TransformCtx,
+    ) -> Result<(), TransformError> {
+        // Perform bottom-up equivalence class analysis.
+        use crate::analysis::DerivedBuilder;
+        let mut builder = DerivedBuilder::new(ctx.features);
+        builder.require(NonNegative);
+        let derived = builder.visit(relation);
+        let derived = derived.as_view();
+
+        self.apply(relation, derived);
+
+        mz_repr::explain::trace_plan(&*relation);
+        Ok(())
+    }
+}
+
+impl WillDistinct {
+    fn apply(&self, expr: &mut MirRelationExpr, derived: DerivedView) {
+        // Record a list of expressions paired with optional "will distinct by" columns.
+        let mut todo = vec![(expr, derived, None::<&[MirScalarExpr]>)];
+        while let Some((expr, derived, distinct_by)) = todo.pop() {
+            // If we find a `Distinct` expression in the shadow of another `Distinct` that will apply to its key columns,
+            // we can remove this `Distinct` operator as the distinctness will be enforced by the other expression.
+            if let (
+                MirRelationExpr::Reduce {
+                    input,
+                    group_key,
+                    aggregates,
+                    ..
+                },
+                Some(columns),
+            ) = (&*expr, distinct_by)
+            {
+                if aggregates.is_empty()
+                    && columns
+                        .iter()
+                        .enumerate()
+                        .all(|(c, e)| e == &MirScalarExpr::Column(c))
+                    && columns.len() == group_key.len()
```

Comment:
This test is more subtle than it should be, in that I got confused by it. The important nuance is that the columns `0 .. k` will correspond to `group_key`, and so if `columns` is exactly those columns, then `columns` and `group_key` correspond. I think.

Reply:
In other words: the `Distinct` operator is both a "distinct" and "project/map" at the same time, and if the outer distinct acts on the projected/mapped columns, we can remove the distincting function of the inner distinct (but still want the map and project behavior).

Reply:
> `columns.len() == group_key.len()`

Would `columns.len() <= group_key.len()` also work? I'm thinking this because when the `Distinct` that is above us is on a subset of the keys of the current `Distinct`, then the above `Distinct` will eliminate _at least_ those rows that the current `Distinct` would also eliminate.

Reply:
Looks like we got the test to something simpler (roughly nothing).